### PR TITLE
Adds more reporting field for cube.

### PR DESCRIPTION
**Pull request by dipak-pawar**

### CODE REVIEW

**Comment by bartoszmajsak** [comment]
Would be great to have some tests.

**Comment by dipak-pawar** [comment]
@bartoszmajsak working on it.

**Comment by bartoszmajsak** [comment]
Cool. I was just thinking maybe in the future we could prefix not finished PRs which you still like to be reviewed as "WIP" and highlight what's still missing in the description?

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/docker/DockerClientExecutor.java [hunk 1]
```diff
@@ -483,6 +484,25 @@ public void startContainer(String id, CubeContainer containerConfiguration) {
...
+    public Statistics statsContainer(String id) {
+         this.readWriteLock.readLock().lock();
+
+        try {
+            StatsCmd statsCmd = this.dockerClient.statsCmd(id);
+            StatsLogsResultCallback statslogs = new StatsLogsResultCallback();
+            try {
+                statsCmd.exec(statslogs).awaitCompletion();
+            } catch (InterruptedException e) {
+                e.printStackTrace();
```
Comment (bartoszmajsak) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/docker/DockerClientExecutor.java]: We should log and throw exception here. Otherwise we can end up with inconsistent state.

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 2]
```diff
@@ -58,9 +75,66 @@ public void reportDockerEnvironment(@Observes AfterAutoStart event, CubeDockerCo
...
+    private void generateStats(DockerClientExecutor executor, Boolean decimal, String when) {
+        if (executor != null) {
+            List<Cube<?>> containers = cubeRegistryInstance.get().getCubes();
 
+            GroupEntry containersStat = new GroupEntry("Container Statistics " + when + " Method");
+
+            for (Cube<?> container : containers) {
+                String name = container.getId();
+                Statistics statistics = executor.statsContainer(name);
+                Map<String, Map> stats = getStats(statistics, decimal);
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: What's the reason behind using raw `Map` type here? Can't we do it more elegant?

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 3]
```diff
@@ -58,9 +75,66 @@ public void reportDockerEnvironment(@Observes AfterAutoStart event, CubeDockerCo
...
+
+        return containerStatsInfo;
     }
 
+   public void reportLogs(@Observes org.arquillian.cube.spi.event.lifecycle.BeforeStop beforeStop, DockerClientExecutor executor, ReporterConfiguration reporterConfiguration) {
+        final String cubeId = beforeStop.getCubeId();
+        if (cubeId != null) {
+            final File logFile = new File(reporterConfiguration.getRootDir(), cubeId + ".log");
+            try {
+                executor.copyLog(beforeStop.getCubeId(), false, true, true, false, -1, new FileOutputStream(logFile));
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: @lordofthejars it looks like executor API needs some love. Lots of magic values here

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 4]
```diff
@@ -58,9 +75,66 @@ public void reportDockerEnvironment(@Observes AfterAutoStart event, CubeDockerCo
...
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            FileEntry fileEntry = new FileEntry();
+            fileEntry.setPath(logFile.getPath());
+            fileEntry.setType(".log");
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: Why `type` is a string? What is it used for?

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java [hunk 5]
```diff
@@ -46,11 +55,23 @@
...
 
+    private CubeRegistry cubeRegistry;
+
+    private Statistics statistics;
+
     @Before
-    public void configureDockerExecutor() {
+    public void beforeTest(){
+        configureDockerExecutor();
+        configuraCube();
```
Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: typo

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java [hunk 6]
```diff
@@ -46,11 +55,23 @@
...
     @Captor
     ArgumentCaptor<PropertyReportEvent> propertyReportEventArgumentCaptor;
 
+    private CubeRegistry cubeRegistry;
+
+    private Statistics statistics;
+
     @Before
-    public void configureDockerExecutor() {
+    public void beforeTest(){
```
Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: Can't we give it a better name?

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 7]
```diff
@@ -263,63 +273,131 @@ private File createSchemasDirectory(File rootDirectory) {
...
+            stats.put("block I/O", blkio);
         }
         return stats;
     }
 
-    private Map<String, String> extractIORW(Map<String, Object> blkioStats, Boolean decimal) {
+    private Map<String, Map<String, String>> extractNetworksStats(Map<String, Object> map, boolean decimal) {
+        Map<String, Map<String, String>> nwStatsForEachNICAndTotal = new LinkedHashMap<>();
+        if (map != null) {
+            long rx_bytes = 0, tx_bytes = 0;
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: I guess, for consistency, we should follow regular java naming conventions.

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java [hunk 8]
```diff
@@ -131,4 +160,48 @@ public void should_report_schema_of_docker_composition() {
         assertThat(screenshotEntry.getLink()).isEqualTo("reports/schemas/docker_composition.png");
     }
 
+   @Test
+    public void shouldRepostLogFile(){
```
Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: We should stick to one test method naming convention. Either camel case or underscore. I would opt for _

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 9]
```diff
@@ -263,63 +273,131 @@ private File createSchemasDirectory(File rootDirectory) {
...
         String pre = (decimal ? "kMGTPE" : "KMGTPE").charAt(exp-1) + (decimal ? "" : "i");
 
         return String.format("%.1f %sB", bytes / Math.pow(unit, exp), pre);
     }
+
+    private long convertToLong(Object number) {
+        long longNumber = 0;
+        String type = number.getClass().getSimpleName();
+        switch (type) {
+            case "Byte":
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: Can't we use enum instead of string for `type`?

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 8 above)
Comment (dipak-pawar) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: ok

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 5 above)
Comment (dipak-pawar) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: Nice catch

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 6 above)
Comment (dipak-pawar) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: sure

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 10]
```diff
@@ -189,8 +279,133 @@ private File createSchemasDirectory(File rootDirectory) {
...
+            long rx_bytes = 0, tx_bytes = 0;
+
+            for (Map.Entry<String, Object> entry: map.entrySet()) {
+                Map<String, String> nwStats = new LinkedHashMap<>();
+                String adapterName = entry.getKey();
+                if (entry.getValue() instanceof LinkedHashMap) {
+
+                    Map<String, ?> adapter = (LinkedHashMap) entry.getValue();
+
+                    long rx_bytes_ = convertToLong(adapter.get("rx_bytes"));
```
Comment (lordofthejars) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: I will change name vars to `rxBytes` and `txBytes` to have a Java like code style.

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java [hunk 11]
```diff
@@ -131,4 +165,48 @@ public void should_report_schema_of_docker_composition() {
...
+
+        assertThat(entryList).hasSize(3).extracting("class.simpleName").containsExactly("GroupEntry", "GroupEntry", "GroupEntry");
+        assertThat(entryList).extracting("name").contains("network statistics", "memory statistics", "block I/O statistics");
+    }
+
+    private void getReportStats(TakeDockerEnvironment takeDockerEnvironment){
+        try {
+            takeDockerEnvironment.reportContainerStatsAfterTest(new After(TakeDockerEnvironmentTest.class, TakeDockerEnvironmentTest.class.getMethod("shouldReportStats")), dockerClientExecutor, cubeRegistry);
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
```
Comment (lordofthejars) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: Avoid using `e.printStackTrace()`

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java [hunk 12]
```diff
@@ -131,4 +165,48 @@ public void should_report_schema_of_docker_composition() {
...
+
+        final PropertyReportEvent propertyReportEvent = propertyReportEventArgumentCaptor.getValue();
+        final PropertyEntry propertyEntry = propertyReportEvent.getPropertyEntry();
+        assertThat(propertyEntry).isInstanceOf(FileEntry.class);
+
+        FileEntry fileEntry = (FileEntry) propertyEntry;
+        assertThat(fileEntry.getPath()).isEqualTo("target/reports/logs/tomcat.log");
+    }
+
+    @Test
```
Comment (lordofthejars) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: I miss another test where you test that it checks the conversion between units.

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 13]
```diff
@@ -58,9 +72,85 @@ public void reportDockerEnvironment(@Observes AfterAutoStart event, CubeDockerCo
...
+
+            GroupEntry containersStat = new GroupEntry("Container Statistics " + when + " Method");
+
+            for (Cube<?> container : containers) {
+                String name = container.getId();
+                Statistics statistics = null;
+                try {
+                    statistics = executor.statsContainer(name);
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: We still have it?

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java [hunk 14]
```diff
@@ -46,11 +56,27 @@
...
+    private Statistics statistics;
+
     @Before
-    public void configureDockerExecutor() {
+    public void configureDockerExecutorAndCubeRegistry(){
+        configureDockerExecutor();
+        try {
+            configureCube();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: Just make this method throwing exception, otherwise it will continue executing tests, where we want to fail fast.

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 12 above)
Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: I would say any logic we introduce should be covered by tests. Ideally test-first. Otherwise how do you know it works? From logs? From user error reports?

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 11 above)
Comment (dipak-pawar) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: you mean we should use logger with some human readbale message

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 14 above)
Comment (dipak-pawar) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: ok

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java (hunk 13 above)
Comment (dipak-pawar) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: I missed to remote it yesterday

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 12 above)
Comment (dipak-pawar) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: @bartoszmajsak , @lordofthejars To construct network Object as I am interested in only rx_bytes & tx_bytes, If I use only those values instead of using all values by keeping schema same. 
e.g. Instead of constructing entire object
`{eth0={rx_bytes=724, rx_packets=7, rx_errors=0, rx_dropped=0, tx_bytes=418, tx_packets=5, tx_errors=0, tx_dropped=0}}`

I can construct only
`{eth0={rx_bytes=724,tx_bytes=418}}`

 WDYT, will this be a good idea?

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/docker/DockerClientExecutor.java [hunk 15]
```diff
@@ -1023,4 +1043,25 @@ public void onNext(Frame object) {
             }
         }
     }
+
+    private static class StatsLogsResultCallback extends ResultCallbackTemplate<StatsLogsResultCallback, Statistics>{
```
Comment (bartoszmajsak) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/docker/DockerClientExecutor.java]: This class is already not only super long, but holds too many responsibilities. Can you extract this embedded one to its own file?

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 16]
```diff
@@ -189,8 +276,136 @@ private File createSchemasDirectory(File rootDirectory) {
...
+            Map<String, String> memory = extractMemoryStats(statistics.getMemoryStats(), decimal, "usage", "max_usage", "limit");
+            Map<String, String> blkio = extractIORW(statistics.getBlkioStats(), decimal);
+            stats.put("network", networks);
+            stats.put("memory", memory);
+            stats.put("block I/O", blkio);
+        }
+        return stats;
+    }
+
+    private Map<String, Map<String, String>> extractNetworksStats(Map<String, Object> map, boolean decimal) {
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: This class has at least 3 responsibilities now

- listens to the Cube events and dispatches some logic
- extract docker-specific information
- does number conversion

I would suggest to split it to three components and do the same for unit tests.

WDYT @lordofthejars @dipak-pawar

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 12 above)
Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: I would go with the smallest usable bits of data for the tests, so the narrowed object sounds good to me, as this would be the only thing we are testing. Is it done already?

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java [hunk 17]
```diff
@@ -131,4 +168,120 @@ public void should_report_schema_of_docker_composition() {
...
+
+        PropertyEntry blkIO = entryList.get(2);
+        assertThat(blkIO).isInstanceOf(GroupEntry.class);
+        List<PropertyEntry> blkIOList = blkIO.getPropertyEntries();
+
+        assertThat(blkIOList).hasSize(2).extracting("class.simpleName").containsExactly("GroupEntry", "GroupEntry");
+        assertThat(blkIOList).flatExtracting("propertyEntries").containsExactly(
+                new KeyValueEntry("I/O Bytes Read", "49.5 KiB"),
+                new KeyValueEntry("I/O Bytes Write", "0 B"));
+    }
```
Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: This test validates 3 functionalities at once. How about 3 separated, focused test methods instead?

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/docker/DockerClientExecutor.java (hunk 15 above)
Comment (dipak-pawar) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/docker/DockerClientExecutor.java]: ok

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 17 above)
Comment (dipak-pawar) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: make sense

Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: I guess it will make our lives easier when it comes to spotting regressions and maintaining test code. Thx for taking care of it.

Comment (dipak-pawar) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: @bartoszmajsak Nopes I have written it. I have used smallest usable bits of data for the tests, except one extra parameter in each object.

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/ExtractContainerStats.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+import com.github.dockerjava.api.model.Statistics;
+import org.arquillian.cube.docker.impl.client.utils.NumberConversion;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ExtractContainerStats {
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/ExtractContainerStats.java]: `ContainerStatsExtractor`?

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/utils/NumberConversion.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
+package org.arquillian.cube.docker.impl.client.utils;
+
+
+import org.arquillian.cube.docker.impl.util.NumberType;
+
+public class NumberConversion {
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/utils/NumberConversion.java]: `NumberConverter` - still too generic, but let's leave it for now

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/utils/NumberConversion.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+    public static String humanReadableByteCount(Long bytes, boolean decimal) {
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/utils/NumberConversion.java]: Can we few unit tests just for this? Some boundary conditions.

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/utils/NumberConversionTest.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+        assertThat(convertToLong((short) 32767), is(32767L));
+        assertThat(convertToLong(100L),is(100L));
+
+        assertThat(convertToLong(2147483647), is(2147483647L));
+        assertThat(convertToLong(0x10), is(16L));
+    }
+
+    @Test
+    public void human_readable_byte_count(){
```
Comment (lordofthejars) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/utils/NumberConversionTest.java]: Last thing before merging. Can you split each assert in its own test? For example in this test you are checking that different conversions are possible. KB, MB, TB, .... so I would write tests like:

```
should_be_possible_to_show_megabytes_in_human_readable_form()
should_be_possible_to_show_gigabytes_in_human_readable_form()
...
```

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/utils/NumberConversionTest.java [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+        assertThat(humanReadableByteCount(null, true), is("0 B"));
+        assertThat(humanReadableByteCount(1024L, true), is("1.02 kB"));
+        assertThat(humanReadableByteCount(1025L, false), is("1.00 KiB"));
+
+        assertThat(humanReadableByteCount(9999800L, true), is("10.00 MB"));
+        assertThat(humanReadableByteCount(281474976710656L,true), is("281.47 TB"));
```
Comment (lordofthejars) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/utils/NumberConversionTest.java]: Can you extract this as constant, so if this represents 10 Megabytes put `private static final Long TEN_MEGABYTES= 9999800`

Comment (dipak-pawar) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/utils/NumberConversionTest.java]: done

**Comment by bartoszmajsak** [comment]
LGTM
